tests/bsim/bluetooth/samples/battery_service/src/peripheral_test.c

```diff
@@ -0,0 +1,187 @@
+
+#define WAIT_TIME 10 /*seconds*/
+
+extern enum bst_result_t bst_result;
+
+static void test_bas_peripheral_init(void)
+{
+	bst_ticker_set_next_tick_absolute(WAIT_TIME * 1e6);
+	bst_result = In_progress;
+}
+
+static void test_bas_peripheral_tick(bs_time_t HW_device_time)
+{
+	/*
+	 * If in WAIT_TIME seconds the testcase did not already pass
+	 * (and finish) we consider it failed
+	 */
+	if (bst_result != Passed) {
+		TEST_FAIL("test_bas_peripheral failed (not passed after %i seconds)\n", WAIT_TIME);
+	}
+}
+
+static const struct bt_data ad[] = {
+	BT_DATA_BYTES(BT_DATA_FLAGS, (BT_LE_AD_GENERAL | BT_LE_AD_NO_BREDR)),
+	BT_DATA_BYTES(BT_DATA_UUID16_ALL, BT_UUID_16_ENCODE(BT_UUID_BAS_VAL)),
+	BT_DATA(BT_DATA_NAME_COMPLETE, CONFIG_BT_DEVICE_NAME, sizeof(CONFIG_BT_DEVICE_NAME) - 1),
+};
+
+static void connected(struct bt_conn *conn, uint8_t err)
+{
+	if (err) {
+		TEST_FAIL("Connection failed (err 0x%02x)\n", err);
+	} else {
+		default_conn = bt_conn_ref(conn);
+
+		LOG_DBG("Peripheral Connected\n");
+	}
+}
+
+static void disconnected(struct bt_conn *conn, uint8_t reason)
+{
+	LOG_DBG("Peripheral %s (reason 0x%02x)\n", __func__, reason);
+
+	if (default_conn) {
+		bt_conn_unref(default_conn);
+		default_conn = NULL;
+	}
+}
+
+static struct bt_conn_cb conn_callbacks = {
+	.connected = connected,
+	.disconnected = disconnected,
+};
+
+static void bt_ready(void)
+{
+	int err;
+
+	LOG_DBG("Peripheral Bluetooth initialized\n");
+
+	err = bt_le_adv_start(BT_LE_ADV_CONN, ad, ARRAY_SIZE(ad), NULL, 0);
+	if (err) {
+		TEST_FAIL("Advertising failed to start (err %d)\n", err);
+		return;
+	}
+
+	LOG_DBG("Advertising successfully started");
+}
+
+static void update_bas_char(void)
+{
+	LOG_DBG("[PERIPHERAL] setting battery level");
+	bt_bas_set_battery_level(90);
+	LOG_DBG("[PERIPHERAL] setting battery present");
+	bt_bas_bls_set_battery_present(BT_BAS_BLS_BATTERY_PRESENT);
+	LOG_DBG("[PERIPHERAL] setting battery charge level");
+	bt_bas_bls_set_battery_charge_level(BT_BAS_BLS_CHARGE_LEVEL_GOOD);
+	LOG_DBG("[PERIPHERAL] setting battery service required true");
+	bt_bas_bls_set_service_required(BT_BAS_BLS_SERVICE_REQUIRED_TRUE);
+}
+
+static void update_bas_char_thread(void *a, void *b, void *c)
+{
+	while (1) {
+		k_sleep(K_SECONDS(1));
+
+		/* Battery level simulation */
+		update_bas_char();
+	}
+}
+
+static void test_bas_peripheral_main(void)
+{
+	int err;
+
+	bt_conn_cb_register(&conn_callbacks);
+
+	/* Mark test as in progress. */
+	TEST_START("peripheral");
+
+	/* Initialize device sync library */
+	bk_sync_init();
+
+	/* Initialize Bluetooth */
+	err = bt_enable(NULL);
+	TEST_ASSERT(err == 0, "Can't enable Bluetooth (err %d)", err);
+
+	LOG_DBG("Bluetooth initialized");
+
+	bt_ready();
+
+	/* Create the thread for updating battery level characteristics */
+	k_thread_create(&update_char_thread, update_char_stack_area,
+			K_THREAD_STACK_SIZEOF(update_char_stack_area), update_bas_char_thread, NULL,
+			NULL, NULL, UPDATE_CHAR_PRIORITY, 0, K_NO_WAIT);
```

Comment:
Isn't it a bit overkill with a thread? Why not just a k_work that is submitted with a 1 second delay?